api/envoy/extensions/filters/http/sxg/v3alpha/sxg.proto

```diff
@@ -0,0 +1,56 @@
+syntax = "proto3";
+
+package envoy.extensions.filters.http.sxg.v3alpha;
+
+import "envoy/extensions/transport_sockets/tls/v3/secret.proto";
+
+import "google/protobuf/duration.proto";
+
+import "udpa/annotations/status.proto";
+
+option java_package = "io.envoyproxy.envoy.extensions.filters.http.sxg.v3alpha";
+option java_outer_classname = "SxgProto";
+option java_multiple_files = true;
+option (udpa.annotations.file_status).work_in_progress = true;
+option (udpa.annotations.file_status).package_version_status = ACTIVE;
+
+// [#protodoc-title: Signed HTTP Exchange Filter]
+// SXG :ref:`configuration overview <config_http_filters_sxg>`.
+// [#extension: envoy.filters.http.sxg]
+
+// [#next-free-field: 10]
+message SXG {
+  // SDS key for x509 certificate data
+  transport_sockets.tls.v3.SdsSecretConfig certificate = 1;
+
+  // SDS key for private key data.
+  transport_sockets.tls.v3.SdsSecretConfig private_key = 2;
+
+  // Duration for which the generated SXG package will be valid. Default is 604800s (7 days in seconds).
```

Comment:
Is that default an SXG recommendation? If so, add link please.

Reply:
@twifkak what do you think? is there any documentation of recommended values for duration?

Reply:
There isn't any canonical documentation for this (it's on my to-do list). My short answer is 1 day if the resource is HTML served without a CSP header or if it is JS (regardless of CSP header), and 7 days otherwise. If differentiating by content type is infeasible, then I would stick to a default of 1 day. But it's not a super strong recommendation.

In your case, I would either a) get rid of the 1-day backdating or b) make the duration 2 days so that it's effectively 1 day from now. I'm torn between these two options; either is fine.

### Details re: duration

What's on my mind is a trade-off between defense-in-depth and speed:

1. If a security vulnerability is discovered in the client-side code, an attacker could ask the server for an SXG and then redistribute it on its own serving path (e.g. via a social media campaign), thereby prolonging the attack for up to `$duration` days after the frontend is fixed.
   - Service workers, having a similar concern, [limit cache duration to 24 hours](https://chromium.googlesource.com/chromium/src/+/main/docs/security/service-worker-security-faq.md#if-a-site-has-an-xss-vulnerability_can-the-attacker-permanently-compromise-that-origin-for-me).
   - JS is the predominant client-side vector for vulnerabilities.
   - CSP headers can mitigate the risks.

2. For SXGs used for prefetching from referrers (the typical use case), a longer duration means a better cache hit rate, which increases its LCP benefit.
   - IME, 1 day is a reasonable "shelf" for the curve of cache hit rate vs duration. Lower than that will cause a more significant drop.

Thus, my default recommendation is 7 day duration for everything except a 1 day duration for:
- JavaScript
- HTML with inline JavaScript without a CSP header

There's logic for that [in AMP Packager](https://github.com/ampproject/amppackager/blob/8ae2d36d8786c5f15c43ca91c5d171532a8185c4/transformer/transformer.go#L288-L315) and in [Web Packager](https://github.com/google/webpackager/blob/6af4d5194ad2f6bc5daf68293f07cc25ba100126/server/from_toml.go#L150-L162) ([here too](https://github.com/google/webpackager/blob/c7597c58f5e83513153ff8b77ad2e7486bd2c597/processor/htmlproc/htmltask/sub_content_type.go#L51-L59)). (The CSP header condition is not implemented; that was a recent idea a colleague came up with.)

Since parsing HTML is probably too heavyweight a dependency for an Envoy extension, I'd assume all HTML contains inline JS, hence my short answer.

### Details re: backdating

 - Option (a) slightly reduces the LCP benefit of prefetching because clients with badly skewed clocks will reject the SXG as too new, but IME the impact of this is <1%, so maybe this an acceptable loss?
 - Option (b) fixes this, but slightly worsens the security posture for those <1% of clients.

Reply:
Interesting. So, perhaps it makes sense to allow for downstream services to specify a duration that the envoy filter can consume. I'll update the docs to suggest 2 days for javascript (1 day, when accounting for the backdating) and 7 days for everything else, with a note about inline JS. Thanks!

Reply:
I can also make hte backdating timespan configurable. let me know what you think makes sense here, and i can update. in the meantime i've updated the notes: ```
  // The duration for which the generated SXG package will be valid. Default is 604800s (7 days in seconds).
  // Note that in order to account for clock skew, the timestamp will be backdated by a day. So, if duration
  // is set to 7 days, that will be 7 days from 24 hours ago (6 days from now). Also note that while 6/7 days 
  // is appropriate for most content, if the downstream service is serving Javascript, or HTML with inline 
  // Javascript, 1 day (so, with backdated expiry, 2 days, or 172800 seconds) is more appropriate.
```

Reply:
The documentation change sounds good (and sufficient) to me.

I'm happy to defer to your opinion on making backdating configurable. I'm hopeful that there's a reasonable default that allows most people not to have to think about it. But if it's cheap to implement and maintain, then it seems nice to have.

For the future version where downstream services specify duration, then maybe something like `min(7 days, [s-maxage or max-age from cache-control header] + $backdate_amount)` would work?